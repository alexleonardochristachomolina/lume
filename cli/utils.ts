import lume from "../mod.ts";
import { toFileUrl } from "../deps/path.ts";
import { isUrl } from "../core/utils/path.ts";
import { getConfigFile } from "../core/utils/lume_config.ts";
import { log } from "../core/utils/log.ts";
import Site from "../core/site.ts";

/** Create a site instance */
export async function createSite(config?: string): Promise<Site> {
  let url: string | undefined;
  const watchInclude = [];

  if (config && isUrl(config)) {
    url = config;
  } else {
    const path = await getConfigFile(config);

    if (path) {
      watchInclude.push(path);
      url = toFileUrl(path).href;
    }
  }

  if (url) {
    log.info(`Loading config file <gray>${url}</gray>`);
    const mod = await import(url);
    const site = mod.default as Site | undefined;
    if (!(site instanceof Site)) {
      log.critical(
        `[Lume] Missing Site instance! Ensure your config file does export the Site instance as default.`,
      );
      throw new Error("Site instance is not found");
    }

    site._data.configFile = url;
    site.options.watcher.include.push(...watchInclude);
    return site;
  }

  return lume();
}

/** Create a site intance and build it */
export async function buildSite(config?: string): Promise<Site> {
  const site = await createSite(config);

  performance.mark("start");
  await site.build();
  performance.mark("end");

  log.info(`üçæ Site built into <gray>${site.options.dest}</gray>`);
  const duration = performance.measure("duration", "start", "end").duration /
    1000;
  const total = site.pages.length + site.files.length;
  log.info(
    `  <gray>${total} files generated in ${duration.toFixed(2)} seconds</gray>`,
  );

  return site;
}
