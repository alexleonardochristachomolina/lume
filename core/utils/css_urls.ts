type UrlWalker = (url: string, type: string) => string | Promise<string>;
type Position = [string, string, number, number];
type Index = [string, number];

export async function walkUrls(
  code: string,
  walker: UrlWalker,
): Promise<string> {
  let result = "";
  let index = 0;
  let position = next(code, index);

  while (position) {
    const [type, url, start, end] = position;
    const newUrl = await walker(url, type);
    result += `${code.slice(index, start)}"${newUrl}"`;
    index = end;
    position = next(code, index);
  }

  return result + code.slice(index);
}

function next(code: string, index = 0): Position | undefined {
  const nextUrl = code.indexOf("url(", index);
  const nextComment = code.indexOf("/*", index);
  const nextImport = code.indexOf("@import ", index);

  let found: Index | undefined;

  if (nextUrl !== -1) {
    found = ["url", nextUrl];
  }
  if (nextComment !== -1 && (!found || nextComment < found[1])) {
    found = ["comment", nextComment];
  }
  if (nextImport !== -1 && (!found || nextImport < found[1])) {
    found = ["import", nextImport];
  }

  if (!found) {
    return;
  }

  // If it's a comment, find the end of it
  if (found[0] === "comment") {
    const end = code.indexOf("*/", found[1]);
    if (end === -1) {
      return;
    }
    return next(code, end + 2);
  }

  if (found[0] === "import") {
    const double = code.indexOf('"', found[1]);
    const single = code.indexOf("'", found[1]);
    const urlFn = code.indexOf("url(", found[1]);

    let foundImport: Index | undefined;

    if (double !== -1) {
      foundImport = ["double", double];
    }
    if (single !== -1 && (!foundImport || single < foundImport[1])) {
      foundImport = ["single", single];
    }

    // It's has a url(), so delegate to the url() logic
    if (urlFn !== -1 && (!foundImport || urlFn < foundImport[1])) {
      return next(code, urlFn);
    }

    if (!foundImport) {
      return;
    }

    const start = foundImport[1];
    const end =
      code.indexOf(foundImport[0] === "double" ? '"' : "'", start + 1) + 1;
    return ["url", code.slice(start + 1, end - 1), start, end];
  }

  if (found[0] === "url") {
    const start = code.indexOf("(", found[1]) + 1;
    const end = code.indexOf(")", start);
    let url = code.slice(start, end).trim();
    // Remove quotes
    if (url.startsWith('"') || url.startsWith("'")) {
      url = url.slice(1, -1);
    }
    return ["url", url, start, end];
  }
}
