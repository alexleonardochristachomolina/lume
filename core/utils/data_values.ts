import type { Data } from "../file.ts";

/**
 * Get the value of a page data
 * For example, if the value is "=title", it returns the value of the page data "title"
 * If the value is "$.title", it will return the value of the element with the selector ".title"
 */
export function getDataValue(data: Partial<Data>, value?: unknown) {
  // Get the value from the page data
  if (typeof value === "string") {
    return searchValue(data, value);
  }

  if (typeof value === "function") {
    return value(data);
  }

  return value;
}

function searchValue(data: Partial<Data>, value: string): unknown {
  if (!value) {
    return;
  }

  if (value.startsWith("=")) {
    let key = value.slice(1);
    [key, value] = parseFallback(key);

    if (!key.includes(".")) {
      return data[key] ?? searchValue(data, value);
    }

    const keys = key.split(".");
    let val = data;
    for (const key of keys) {
      val = val?.[key];
    }
    return val ?? searchValue(data, value);
  }

  if (value.startsWith("$")) {
    let selector = value.slice(1);
    [selector, value] = parseFallback(selector);

    return queryCss(selector, data.page?.document) ?? searchValue(data, value);
  }

  return value;
}

function parseFallback(key: string): [string, string] {
  const fallback = key.indexOf("||");

  if (fallback !== -1) {
    return [
      key.slice(0, fallback).trim(),
      key.slice(fallback + 2).trim(),
    ];
  }

  return [
    key,
    "",
  ];
}

// https://regexr.com/7qnot
const checkForAttrPattern = /^(.+)\s+(?:attr\(([\w\-]+)\))$/;

function queryCss(query: string, document?: Document) {
  const checkResult = query.match(checkForAttrPattern);

  const hasAttr = checkResult?.[0];
  if (hasAttr) {
    const [_, query, name] = checkResult;
    return document?.querySelector(query)?.getAttribute(name);
  }

  return document?.querySelector(query)?.innerHTML;
}
