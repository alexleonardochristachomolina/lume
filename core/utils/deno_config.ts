import { parse } from "../../deps/jsonc.ts";
import { isUrl } from "../utils/path.ts";

/** Import map file */
export interface ImportMap {
  imports: Record<string, string>;
  scopes?: Record<string, Record<string, string>>;
}

/** Basic options for deno.json file */
export interface DenoConfig extends Partial<ImportMap> {
  importMap?: string;
  tasks?: Record<string, string>;
  compilerOptions?: {
    jsx?: "jsx" | "react-jsx" | "precompile";
    jsxImportSource?: string;
    types?: string[];
  };
  [key: string]: unknown;
}

export interface DenoConfigResult {
  file: string;
  config: DenoConfig;
  importMap?: ImportMap;
}

/** Detect and returns the Deno configuration */
export async function readDenoConfig(): Promise<DenoConfigResult | undefined> {
  for (const file of ["deno.json", "deno.jsonc"]) {
    try {
      const content = await Deno.readTextFile(file);
      const config = parse(content) as DenoConfig;
      let importMap: ImportMap | undefined;

      if (config.importMap) {
        importMap = isUrl(config.importMap)
          ? await (await fetch(config.importMap)).json()
          : await JSON.parse(await Deno.readTextFile(config.importMap));
      } else if (config.imports) {
        importMap = {
          imports: config.imports,
          scopes: config.scopes,
        };
      }
      return { file, config, importMap };
    } catch (err) {
      if (err instanceof Deno.errors.NotFound) {
        continue;
      }

      throw err;
    }
  }
}
