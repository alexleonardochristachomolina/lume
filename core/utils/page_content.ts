import type { Page } from "../file.ts";

export function insertContent(
  page: Page,
  content: string,
  placeholder?: string,
) {
  const pageContent = page.content;

  if (pageContent && typeof pageContent === "string") {
    if (placeholder && pageContent.includes(placeholder)) {
      page.content = pageContent.replace(
        placeholder,
        content,
      );
    } else {
      page.content += `\n${content}`;
    }
  } else {
    page.content = content;
  }
}
