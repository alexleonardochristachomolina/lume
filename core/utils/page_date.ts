import { getGitDate, parseDate } from "./date.ts";

import type { Page, RawData } from "../file.ts";

/** Returns the Date instance of a file */
export function getPageDate(page: Page): Date {
  const data = page.data as RawData;
  const { date } = data;

  if (date instanceof Date) {
    return date;
  }

  if (typeof date === "number") {
    return new Date(date);
  }

  const { entry } = page.src;
  const info = entry?.getInfo();

  if (typeof date === "string") {
    if (entry && info) {
      switch (date.toLowerCase()) {
        case "git last modified":
          return getGitDate("modified", entry.src) || info.mtime || new Date();
        case "git created":
          // getGitDate("created", ...) uses `git log --diff-filter=A`
          // which returns nothing for files where the latest commit had a merge conflict.
          // Therefore we fallback to last modified time from git history,
          // which is more reliable than file system creation time under most CI environments.
          return getGitDate("created", entry.src) ||
            getGitDate("modified", entry.src) || info.birthtime ||
            new Date();
      }
    }

    try {
      return parseDate(date);
    } catch {
      throw new Error(`Invalid date: ${date} (${entry?.src})`);
    }
  }

  return info?.birthtime || info?.mtime || new Date();
}
