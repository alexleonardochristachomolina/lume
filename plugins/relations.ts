import { merge } from "../core/utils/object.ts";

import type Site from "../core/site.ts";
import type { Data, Page } from "../core/file.ts";

type RelationFilter = (data1: Data, data2: Data) => boolean;

interface ForeignKeyOptions {
  foreignKey: string;
  relationKey?: string;
  pluralRelationKey?: string;
  idKey?: string;
  filter?: RelationFilter;
}

export interface Options {
  /** The list of extensions this plugin applies to */
  extensions?: string[];

  /** The field name used to save the page id */
  idKey?: string;

  /** The field name used to save the page type */
  typeKey?: string;

  /** The foreign keys per type (type => foreign_key) */
  foreignKeys: Record<string, string | ForeignKeyOptions>;
}

// Default options
export const defaults: Options = {
  extensions: [".html"],
  idKey: "id",
  typeKey: "type",
  foreignKeys: {},
};

/**
 * A plugin to create relations between pages
 * @see https://lume.land/plugins/relations/
 */
export function relations(userOptions: Options) {
  const options = merge(defaults, userOptions);

  return (site: Site) => {
    site.preprocess(options.extensions, index);

    function index(pages1: Page[], pages: Page[]) {
      pages1.forEach((page1) => {
        const data1 = page1.data;
        const [
          type1,
          foreignKey1,
          id1,
          relationType1,
          pluralRelationKey1,
          filter1,
        ] = getRelationInfo(data1);

        // Index the current page with the other pages
        pages.forEach(indexPage);

        // Index the current page with previously generated pages (if any)
        site.pages.forEach(indexPage);

        function indexPage(page2: Page) {
          if (page1 === page2) {
            return;
          }

          const data2 = page2.data;

          if (filter1 && !filter1(data1, page2.data)) {
            return;
          }

          const [
            type2,
            foreignKey2,
            id2,
            relationType2,
            pluralRelationKey2,
            filter2,
          ] = getRelationInfo(data2);

          if (filter2 && !filter2(data2, page1.data)) {
            return;
          }

          // Page2 has a foreign key to page1
          const directRelation = relate(
            data1,
            data2,
            foreignKey1,
            id1,
            type1,
            relationType1,
            pluralRelationKey1,
          );
          // If it was related, do the opposite relation
          if (directRelation && pluralRelationKey2) {
            saveMultipleRelation(
              data2,
              data1,
              pluralRelationKey2,
            );
            return;
          }

          // Page1 has a foreign key to page2
          const reverseRelation = relate(
            data2,
            data1,
            foreignKey2,
            id2,
            type2,
            relationType2,
            pluralRelationKey2,
          );

          // If it was related, do the opposite relation
          if (reverseRelation && pluralRelationKey1) {
            saveMultipleRelation(
              data1,
              data2,
              pluralRelationKey1,
            );
          }
        }

        function relate(
          rel: Data,
          data: Data,
          foreignKey?: string,
          id?: string,
          type?: string,
          relationKey?: string,
          pluralRelationKey?: string,
        ): boolean {
          if (foreignKey && type && id && data[foreignKey]) {
            const relId = data[foreignKey] as string | string[];

            // The foreign key contain an array
            if (Array.isArray(relId)) {
              if (relId.includes(id)) {
                saveMultipleRelation(rel, data, pluralRelationKey || type);
                return true;
              }
              return false;
            }

            // The foreign key is a single value
            if (relId == id) {
              data[relationKey || type] = rel;
              return true;
            }
          }

          return false;
        }

        function saveMultipleRelation(rel: Data, data: Data, type: string) {
          const relData = (data[type] || []) as Data[];

          if (!relData.includes(rel)) {
            relData.push(rel);
            data[type] = relData;
            // Sort by id
            relData.sort((a, b) => {
              const idA = a[options.idKey] as string;
              const idB = b[options.idKey] as string;
              return idA < idB ? -1 : 1;
            });
          }
        }
      });
    }
  };

  function getRelationInfo(
    data: Data,
  ): [string?, string?, string?, string?, string?, RelationFilter?] {
    const type = data[options.typeKey];
    if (!type) {
      return [];
    }

    const foreignKey = options.foreignKeys[type];
    if (!foreignKey) {
      return [type, undefined, undefined, type, type];
    }

    if (typeof foreignKey === "string") {
      return [type, foreignKey, data[options.idKey], type, type];
    }

    return [
      type,
      foreignKey.foreignKey,
      data[foreignKey.idKey || options.idKey],
      foreignKey.relationKey || type,
      foreignKey.pluralRelationKey || foreignKey.relationKey || type,
      foreignKey.filter,
    ];
  }
}

export default relations;
