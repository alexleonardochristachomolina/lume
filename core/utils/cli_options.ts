import { parseArgs } from "../../deps/cli.ts";
import type { DeepPartial } from "./object.ts";
import type { SiteOptions } from "../site.ts";

export function getOptionsFromCli(
  options: DeepPartial<SiteOptions>,
): DeepPartial<SiteOptions> {
  const cli = parseArgs(Deno.args, {
    string: ["src", "dest", "location", "port"],
    boolean: ["serve", "open"],
    alias: { dev: "d", serve: "s", port: "p", open: "o" },
    ["--"]: true,
  });

  if (cli.src) {
    options.src = cli.src;
  }

  if (cli.dest) {
    options.dest = cli.dest;
  }

  const serveMode = cli.serve || cli._[0] === "cms";

  if (cli.port) {
    (options.server ||= {}).port = parseInt(cli.port);
  } else if (serveMode) {
    (options.server ||= {}).port = 3000;
  }

  let location: URL;

  if (cli.location) {
    location = new URL(cli.location);
  } else if (options.location && !serveMode) {
    location = options.location as URL;
  } else {
    location = new URL("http://localhost");
  }

  let port: number;

  if (cli.port) {
    port = parseInt(cli.port);
  } else if (location.port) {
    port = parseInt(location.port);
  } else if (options.server?.port) {
    port = options.server.port;
  } else {
    port = cli.serve ? 3000 : location.protocol === "https:" ? 443 : 80;
  }

  (options.server ||= {}).port = port;
  location.port = port.toString();
  options.location = location;

  if (cli.open) {
    (options.server ||= {}).open = cli.open;
  }

  return options;
}
