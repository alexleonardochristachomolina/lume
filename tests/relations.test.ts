import { assertSiteSnapshot, build, getSite } from "./utils.ts";
import relations from "../plugins/relations.ts";

Deno.test("relations plugin", async (t) => {
  const site = getSite({
    src: "relations",
  });

  site.use(relations({
    foreignKeys: {
      post: "post_id",
      category: { foreignKey: "category_id", idKey: "basename" },
      comment: { foreignKey: "comment_id", pluralRelationKey: "comments" },
    },
  }));

  await build(site);
  await assertSiteSnapshot(t, site);
});
