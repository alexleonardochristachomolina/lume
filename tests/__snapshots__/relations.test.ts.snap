export const snapshot = {};

snapshot[`relations plugin 1`] = `
{
  formats: [
    {
      engines: 0,
      ext: ".page.toml",
      loader: [AsyncFunction: toml],
      pageType: "page",
    },
    {
      engines: 1,
      ext: ".page.ts",
      loader: [AsyncFunction: module],
      pageType: "page",
    },
    {
      engines: 1,
      ext: ".page.js",
      loader: [AsyncFunction: module],
      pageType: "page",
    },
    {
      engines: 0,
      ext: ".page.jsonc",
      loader: [AsyncFunction: json],
      pageType: "page",
    },
    {
      engines: 0,
      ext: ".page.json",
      loader: [AsyncFunction: json],
      pageType: "page",
    },
    {
      dataLoader: [AsyncFunction: json],
      engines: 0,
      ext: ".json",
      loader: [AsyncFunction: json],
    },
    {
      dataLoader: [AsyncFunction: json],
      engines: 0,
      ext: ".jsonc",
      loader: [AsyncFunction: json],
    },
    {
      engines: 1,
      ext: ".md",
      loader: [AsyncFunction: text],
      pageType: "page",
    },
    {
      engines: 1,
      ext: ".markdown",
      loader: [AsyncFunction: text],
      pageType: "page",
    },
    {
      dataLoader: [AsyncFunction: module],
      engines: 1,
      ext: ".js",
      loader: [AsyncFunction: module],
    },
    {
      dataLoader: [AsyncFunction: module],
      engines: 1,
      ext: ".ts",
      loader: [AsyncFunction: module],
    },
    {
      engines: 1,
      ext: ".vento",
      loader: [AsyncFunction: text],
      pageType: "page",
    },
    {
      engines: 1,
      ext: ".vto",
      loader: [AsyncFunction: text],
      pageType: "page",
    },
    {
      dataLoader: [AsyncFunction: toml],
      engines: 0,
      ext: ".toml",
      loader: [AsyncFunction: toml],
    },
    {
      dataLoader: [AsyncFunction: yaml],
      engines: 0,
      ext: ".yaml",
      loader: [AsyncFunction: yaml],
      pageType: "page",
    },
    {
      dataLoader: [AsyncFunction: yaml],
      engines: 0,
      ext: ".yml",
      loader: [AsyncFunction: yaml],
      pageType: "page",
    },
  ],
  src: [
    "/",
    "/_includes",
    "/_includes/categories.vto",
    "/_includes/comment.vto",
    "/_includes/posts.vto",
    "/categories",
    "/categories/_data.yml",
    "/categories/category-1.md",
    "/categories/category-2.md",
    "/comments",
    "/comments/_data.yml",
    "/comments/comment-1.md",
    "/comments/comment-2.md",
    "/comments/comment-3.md",
    "/posts",
    "/posts/_data.yml",
    "/posts/post-1.md",
    "/posts/post-2.md",
  ],
}
`;

snapshot[`relations plugin 2`] = `[]`;

snapshot[`relations plugin 3`] = `
[
  {
    content: "<!DOCTYPE html>
<h1>Category 1</h1>

<div><p>This is the first category</p>
</div>

<h2>Posts</h2>
<ul>

</ul>
",
    data: {
      basename: "category-1",
      children: "<p>This is the first category</p>
",
      content: "This is the first category
",
      date: [],
      layout: "categories.vto",
      mergedKeys: [
        "tags",
      ],
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      post: "Array(1)",
      renderOrder: -2,
      search: [],
      tags: "Array(0)",
      title: "Category 1",
      type: "category",
      url: "/categories/category-1/",
    },
    src: {
      asset: false,
      ext: ".md",
      path: "/categories/category-1",
      remote: undefined,
    },
  },
  {
    content: "<!DOCTYPE html>
<h1>Category 2</h1>

<div><p>This is the second category</p>
</div>

<h2>Posts</h2>
<ul>

</ul>
",
    data: {
      basename: "category-2",
      children: "<p>This is the second category</p>
",
      content: "This is the second category
",
      date: [],
      layout: "categories.vto",
      mergedKeys: [
        "tags",
      ],
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      post: "Array(1)",
      search: [],
      tags: "Array(0)",
      title: "Category 2",
      type: "category",
      url: "/categories/category-2/",
    },
    src: {
      asset: false,
      ext: ".md",
      path: "/categories/category-2",
      remote: undefined,
    },
  },
  {
    content: "<!DOCTYPE html>
<p>This is the first comment</p>
",
    data: {
      basename: "comment-1",
      children: "<p>This is the first comment</p>
",
      content: "This is the first comment
",
      date: [],
      id: 1,
      mergedKeys: [
        "tags",
      ],
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      post: "Array(1)",
      search: [],
      tags: "Array(0)",
      type: "comment",
      url: "/comments/comment-1/",
    },
    src: {
      asset: false,
      ext: ".md",
      path: "/comments/comment-1",
      remote: undefined,
    },
  },
  {
    content: "<!DOCTYPE html>
<p>This is the second comment</p>
",
    data: {
      basename: "comment-2",
      children: "<p>This is the second comment</p>
",
      content: "This is the second comment
",
      date: [],
      id: 2,
      mergedKeys: [
        "tags",
      ],
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      post: "Array(1)",
      renderOrder: -1,
      search: [],
      tags: "Array(0)",
      type: "comment",
      url: "/comments/comment-2/",
    },
    src: {
      asset: false,
      ext: ".md",
      path: "/comments/comment-2",
      remote: undefined,
    },
  },
  {
    content: "<!DOCTYPE html>
<p>This is the third comment</p>
",
    data: {
      basename: "comment-3",
      children: "<p>This is the third comment</p>
",
      content: "This is the third comment
",
      date: [],
      id: 3,
      mergedKeys: [
        "tags",
      ],
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      post: "Array(1)",
      search: [],
      tags: "Array(0)",
      type: "comment",
      url: "/comments/comment-3/",
    },
    src: {
      asset: false,
      ext: ".md",
      path: "/comments/comment-3",
      remote: undefined,
    },
  },
  {
    content: '<!DOCTYPE html>
<div><p>First post with category 2 and comments 1 and 3</p>
</div>


<h2>Category:</h2>
<a href="/categories/category-2/">Category 2</a>
<p>This is the second category</p>


<ul>

</ul>
',
    data: {
      basename: "post-1",
      category: [
        "basename",
        "mergedKeys",
        "paginate",
        "search",
        "tags",
        "type",
        "layout",
        "title",
        "content",
        "url",
        "date",
        "page",
        "children",
        "post",
      ],
      category_id: "category-2",
      children: "<p>First post with category 2 and comments 1 and 3</p>
",
      comment_id: "Array(2)",
      comments: "Array(2)",
      content: "First post with category 2 and comments 1 and 3
",
      date: [],
      id: "first-post",
      layout: "posts.vto",
      mergedKeys: [
        "tags",
      ],
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      renderOrder: 1,
      search: [],
      tags: "Array(0)",
      type: "post",
      url: "/posts/post-1/",
    },
    src: {
      asset: false,
      ext: ".md",
      path: "/posts/post-1",
      remote: undefined,
    },
  },
  {
    content: '<!DOCTYPE html>
<div><p>Second post with category 1 and comment 2</p>
</div>


<h2>Category:</h2>
<a href="/categories/category-1/">Category 1</a>
<p>This is the first category</p>


<ul>

</ul>
',
    data: {
      basename: "post-2",
      category: [
        "basename",
        "mergedKeys",
        "paginate",
        "search",
        "tags",
        "type",
        "layout",
        "title",
        "renderOrder",
        "content",
        "url",
        "date",
        "page",
        "children",
        "post",
      ],
      category_id: "category-1",
      children: "<p>Second post with category 1 and comment 2</p>
",
      comment_id: "Array(1)",
      comments: "Array(1)",
      content: "Second post with category 1 and comment 2
",
      date: [],
      id: "second-post",
      layout: "posts.vto",
      mergedKeys: [
        "tags",
      ],
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      search: [],
      tags: "Array(0)",
      type: "post",
      url: "/posts/post-2/",
    },
    src: {
      asset: false,
      ext: ".md",
      path: "/posts/post-2",
      remote: undefined,
    },
  },
]
`;
